Create JB blog (squashed) (#5)
mkondratek/git-machete-blog-post: blog.md

## Take a look at your repository from a new perspective with Git Machete Plugin!

Git and IntelliJ-based software are currently one of the most important tools in software development.
The version control system provided with the JetBrains's IDE products
allows performing pretty much all everyday VCS actions directly via the IDE.
Among the many useful benefits of the provided IDE and VCS integration,

> the full support of GitHub Pull Requests introduced in IDEA 2020.2 must be emphasized.

Reviewer: `s/emphasized/highlighted`? :thinking:  The sentence sounds too stiff/formal

It finds an outstanding application in maintaining linear Git history.

## Git Machete

The Git Machete plugin is a tool that helps developers handle multiple branches.

> The main features of the plugin are:

Reviewer: I wouldn't call the below `features`, rather more high-level `objectives`? `aims`?...

Reviewer: Also, `s/are/include` sounds better

- helps keep branches up to date even if (side effects of small PRs):
  - there many of them

>   - they're occasionally stacked on one another

Reviewer: ```suggestion
  - they're occasionally stacked on one another, e.g. `develop` (arrow) `refactor/foo` (arrow) `bugfix/bar`
```

- promotes linear history but without strictly enforcing it in every case
- gives a bird's eye view of the repo
- automates common flows...

> #### Git Machete Status Graph

Reviewer: Hmm fourth-level header under the second-level one?

![Git Machete - Actions](actions-after_dark.png "Git Machete - Actions - After")

### Motivation

The idea standing behind the plugin is automating the developer's work.

> We (developers) should strive for focusing on actual development

Reviewer: `developX` repeated  3 times :c

and spend less time organizing our work (with no loss for its results!).

To be more precise and understand the technical background
we should look closer to the consequences of big and small PRs.

> **Big PRs are bad.** Reviewers are unwilling to even begin the review.

Reviewer: ```suggestion
**Big PRs are bad,** especially if they mix refactors with bugfixes and new features.
Reviewers are less willing to even begin the review.
```

They are more likely to skip vital parts of the newly-added code.
It's hard/impossible to keep entire context in memory during review.

On the contrary, **small PRs are beneficial**.
Reviewers are more willing to begin the review.
They are less likely to skip vital parts of the newly-added code.
It's much easier to keep the entire context in memory during the review.

Creating and maintaining **small, focused, easy-to-review PRs** is a comfortable and desired approach
for both the authors and, especially, the reviewers.
This is exactly what Git Machete is promoting.

### The future of Git Machete

Git Machete plugin is being developed and maintained continuously.
We are doing our best to provide new features and quickly react to the users' feedback.
We are responding to the new issues and ideas.

> The feature that we are definitely aiming for now is the traverse functionality

Reviewer: ```suggestion
The feature that we are definitely aiming for now is the semi-automatic traversal functionality
```

![Git Machete - Actions](actions-after_dark.png "Git Machete - Actions - After")

## Motivation

The idea standing behind the plugin is automating the developer's work.
One should strive for focusing on actual software engineering
and spend less time organizing our work (with no loss for its results!).

To be more precise and understand the technical background
we should look closer to the consequences of big and small PRs.
**Big PRs are bad,** especially if they mix refactors with bugfixes and new features.
Reviewers are unwilling to even begin the review.
They are more likely to skip vital parts of the newly-added code.

> It's hard/impossible to keep entire context in memory during review.

Reviewer 2: ```suggestion
It's hard/impossible to keep the entire context in memory during a review.
```

On the contrary, **small PRs are beneficial**.
Reviewers are more willing to begin the review.
They are less likely to skip vital parts of the newly-added code.
It's much easier to keep the entire context in memory during the review.

Creating and maintaining **small, focused, easy-to-review PRs** is a comfortable and desired approach
for both the authors and, especially, the reviewers.

> This is exactly what Git Machete is promoting.

Reviewer 2: ```suggestion
That is exactly what Git Machete is promoting.
```

## The future of Git Machete

Git Machete plugin is being developed and maintained continuously.
We are doing our best to provide new features and quickly react to the users' feedback.

> We are responding to the new issues and ideas.

Reviewer 2: ```suggestion
We are responding to new issues and ideas.
```

The feature that we are definitely aiming for now is the semi-automatic traversal functionality
which would allow you to interactively go through the whole branch layout tree
and apply necessary actions (rebases, pushes, etc.) in order to get every branch in syncs.

## Learn more

The plugin is available from the [Marketplace](https://plugins.jetbrains.com/plugin/14221-git-machete "JetBrains Marketplace - Git Machete Plugin").

> It may differ a bit from presented version as some improvements may had been provided with the newest release.

Reviewer 2: ```suggestion
It may differ a bit from the presented version as some improvements may have been provided with the newest release.
```
